### PR TITLE
QuickNav Text paragraph navigation

### LINKED ISSUE #15998: Feature request: add QuickNav command to jump to next/previous text paragraph
I propose to add  QuickNav command to jump to next/previous text paragraph to NVDA. I I will implement this feature request myself if green light is given by NVDA team.

* Here we define text paragraph as paragraph with at least one sentence, that is written with punctuation marks, such as period, question mark, exclamation mark, their full-width CJK counterparts, etc.
* This allows users to quickly skip over clutter (such as menus, ads, etc) on a wide variety of websites.
* Another illustration of its utility is finding the beginning of an article on an unfamiliar news web site.
* This function is a heuristic that doesn't work 100% of the time. There are ads written in complete sentences, and likewise, sometimes in readable text of an article there can be a paragraph without any punctuation marks that will be missed. However, in practice this function works very well most of the time on a wide range of websites and simplifies navigation significantly.
* This feature has already been implemented in [TextNav add-on](https://addons.nvda-project.org/addons/textnav.en.html) and by default assigned to `Alt+Shift+Up/Down` gestures.
* The add-on is quite stable, it has been around for 5 years and for the last 4 years it's been in maintenance mode and didn't require any fixing besides NVDA compatibility updates.
* TextNav add-on is quite popular among NVDA users and has been frequently recommended on NVDA mailing list. Recently four users have conteacted me and asked to make it compatible with NVDA v2024.1. Moreover, a quick search reveals that people recommend TextNav in their blogs:
    * https://webfriendlyhelp.com/textnav/
    * (Misspelled as TestNav) https://slnconline.com.ng/10_nvda_addon_to_boost_your_productivity_and_effeciency.html
* In spirit this feature is similar to built-in N QuickNav command "Skip forward past a block of links", which in my understanding was also designed to skip menus on the websites. However in my experience jumping to text paragraph works much better then N QuickNav command as it works more reliably and typically finds exactly the first paragraph of the article.
* Finally, I believe this command will be quite useful for NVDA beginners, as one of the top tasks they struggle with is trying to make sense of web pages while trying to memorize dozens of keystrokes. With this command they'll be able to quickly and easily start using news articles, which I would think is one of the major use cases for the beginners.
* I propose to assign by default  QuickNav P and Shift+P gestures to this command.

### PR DESCRIPTION
**Pull request by mltony**
### Link to issue number:
Closes #15998
### Summary of the issue:
Add QuickNav text paragraph navigation
### Description of user facing changes
* Added QuickNav gesture `p` to jump to next/previous text paragraph in browse mode.
### Description of development approach
* In `BrowseModeTreeInterceptor` added function `_iterSimilarParagraph` that finds next/previous paragraph that satisfies condition defined by a lambda function. I will reuse this function in later PRs to implement vertical navigation.
* Added a new clause in `BrowseModeTreeInterceptor._quickNavScript()` that handles text paragraph case and calls function defined in the previous bullet.
* For text criteria I ended up implementing a user configurable regex. 
    * Its initial value is defined in `DEFAULT_TEXT_PARAGRAPH_REGEX` variable in `configSpec.py:11`. 
    * It is user-configurable in Browse Mode page in NVDA options.
    * The rationale for using a regex instead of plain-text search is to reduce the number of false positives.
    * For example, for period character we check that it follows a word character `\w` and is followed by space character or end of string. There are a few more rules for edge cases in the code.
    * I don't include comma, colon and semicolon punctuation marks in the regex due to high number of false positives.
    * The regex also accounts for CJK languages by checking for full-width punctuation marks.
### Testing strategy:
* Tested manually on a bunch of webpages in Google Chrome and Mozilla Firefox.
* Included a system test.
### Known issues with pull request:
N/A
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

**Comment by mltony** [comment]
Also looking for feedback from the international community on how does this PR do with your languages. I'll leave this PR as a draft for a few days to collect and address this feedback. Specifically, tell us how does this PR work with your language and also would be great to hear punctuation rules in your language if different from Latin. Here are major languages that I'm interested in:
* CJK languages. I included Full-width CJK punctuation marks but I don't speak any CJK languages, so wondering if any extra conditions need to be added  for CJK clause.
* Languages written in Arabic script - Arabic, Farsi.
* Languages of India.
Also feel free to provide feedback for any other languages you speak.

File: source/config/configSpec.py [hunk 1]
```diff
@@ -8,6 +8,31 @@
...
+		punc=r'[.!?…]{1,3}',
+		# On Wikipedia references appear right after period in sentences, the following clause takes this
+		# into account. For example:
+		# > On his father's side, he was a direct descendant of John Churchill.[3]
+		optWiki=r'(\[\d+\])*',
+		# LookAhead clause checks that punctuation mark must be followed by either space,
+		# or newLine symbol or end of string.
+		lookAhead=r'(?=[\r\n  ]|$)',
+		# We also check for CJK full-width punctuation marks without any extra rules.
+		cjk=r'[．！？：；]',
```
Comment (hwf1324) [inline source/config/configSpec.py:32]:
```suggestion
		cjk=r'[。．！？：；]',

```

For Chinese "。" is used a bit more.

File: tests/system/robot/chromeTests.py [hunk 2]
```diff
@@ -2473,3 +2473,48 @@ def test_i13307():
...
+		<p>test....</p>
+		<p>a.b</p>
+		<p></p>
+		<!-- Now a bunch of matching paragraphs -->
+		<p>Hello, world!</p>
+		<p>He replied, "That's wonderful."</p>
+		<p>He replied, "That's wonderful".</p>
+		<p>He replied, "That's wonderful."[4]</p>
+		<p>Предложение по-русски.</p>
+		<p>我不会说中文！</p>
```
Comment (hwf1324) [inline tests/system/robot/chromeTests.py:2497]:
```suggestion
		<p>这是中文段落。</p>
		<p>我不会说中文！</p>
```

File: tests/system/robot/chromeTests.py [hunk 3]
```diff
@@ -2473,3 +2473,48 @@ def test_i13307():
...
+		<p>Bye-bye, world!</p>
+	""")
+
+	expectedParagraphs = [
+		'''Hello, world!''',
+		'''He replied,  That's wonderful.''',
+		'''He replied,  That's wonderful .''',
+		'''He replied,  That's wonderful.  4''',
+		'''Предложение по-русски.''',
+		'''我不会说中文！''',
```
Comment (hwf1324) [inline tests/system/robot/chromeTests.py:2507]:
```suggestion
		'''这是中文段落。''',
		'''我不会说中文！''',
```

**Review by hwf1324 — CHANGES_REQUESTED** [review]

**Comment by Adriani90** [comment]
cc: @cary-rowen, @nishimotz, @dpy013, @josephsl for chinese, japanese and corean at least.

**Comment by amirsol81** [comment]
@mltony P and SHIFT+P conflict with the go to next/previous dialog commands in your BrowserNav add-on. Any chance of also updating BrowserNav?

**Comment by mltony** [comment]
@amirsol81, Right, I will soon remove P command from BrowserNav to avoid keystroke collisions. You should use built-in QuickNav O instead of BrowserNav P which does the same thing.

File: source/browseMode.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 from typing import (
 	Any,
 	Callable,
+	Generator,
```
Comment (lukaszgo1) [inline source/browseMode.py:10]: We are on Python 3.11, so this should use `from collections.abc import Generator`

File: source/browseMode.py [hunk 5]
```diff
@@ -440,9 +442,53 @@ def _iterNodesByType(self,itemType,direction="next",pos=None):
 	def _iterNotLinkBlock(self, direction="next", pos=None):
 		raise NotImplementedError
 
+	def _iterSimilarParagraph(
+			self,
+			kind: str,
+			paragraphFunction: Callable[[textInfos.TextInfo], Optional[Any]],
+			desiredValue: Optional[Any] = None,
+			direction: str = "next",
+			pos: textInfos.TextInfo = None,
```
Comment (lukaszgo1) [inline source/browseMode.py:451]: `None` is not a valid value for this parameter, and having a incorrect default here means someone may call this method without specifying `pos`. Can params be reordered here so that specifying incorrect default is not necessary?

File: source/browseMode.py [hunk 6]
```diff
@@ -440,9 +442,53 @@ def _iterNodesByType(self,itemType,direction="next",pos=None):
...
+				yield TextInfoQuickNavItem(kind, self, info.copy())
+
+
 	def _quickNavScript(self,gesture, itemType, direction, errorMessage, readUnit):
 		if itemType=="notLinkBlock":
 			iterFactory=self._iterNotLinkBlock
+		elif itemType == "textParagraph":
+			punctuationMarksRegex = re.compile(
+				config.conf["virtualBuffers"]["textParagraphRegex"],
+				re.UNICODE,
```
Comment (lukaszgo1) [inline source/browseMode.py:478]: `re.UNICODE` is the default in Python 3, so it can be omitted. The same goes for the second usage further down.

File: source/gui/settingsDialogs.py [hunk 7]
```diff
@@ -2272,6 +2272,36 @@ def makeSettings(self, settingsSizer):
...
+			log.debugWarning("Failed to compile text paragraph regex", exc_info=True)
+			gui.messageBox(
+				# Translators: Message shown when invalid text paragraph regex entered
+				_("Failed to compile text paragraph regular expression: %s") % str(e),
+				# Translators: The title of the message box
+				_("Error"),
+				wx.OK | wx.ICON_ERROR,
+				self,
+			)
+			self.textParagraphRegexEdit.SetFocus()
```
Comment (lukaszgo1) [inline source/gui/settingsDialogs.py:2300]: Setting focus when validating should be avoided, as it behaves unexpectedly when applying settings for a different panel (see #15894 for more discussion).

File: source/browseMode.py (hunk 4 above)
Comment (mltony) [reply source/browseMode.py:10]: [Here](https://www.phind.com/search?cache=q2x75q8negs9q9yv5zjrh8tx) they say `collections.abc.Generator` use for type hinting is deprecated in favor of `typing.Generator`. And I havent found any usage of `abc.Generator` as type hint in NVDA. Do you have any evidence to support your point of view?

Comment (seanbudd) [reply source/browseMode.py:10]: The docs themselves say the deprecation is the other way around - that `typing.Generator` is the deprecated alias
https://docs.python.org/3/library/typing.html#typing.Generator

File: source/browseMode.py (hunk 5 above)
Comment (mltony) [reply source/browseMode.py:451]: The way I am using this function is called partial application - this term is from functional programming, not sure if it is used in Pythonic context, but basically in `_quickNavScript` I only resolve the first two function arguments and return the function of the remaining arguements. Even though the order is flexible, it is considered to be a good style to declare function arguments in the order they will be resolved. Moreover pos seems to be the last argument in other similar functions. So But I can remove default values from all arguments.

**Comment by nishimotz** [comment]
The Windows Uniscribe API is used for multilingual processing of word and sentence segmentation.
It is likely already being utilized in NVDA's textInfos.
This API might also be applicable for paragraph processing.

File: tests/system/robot/chromeTests.py [hunk 8]
```diff
@@ -2473,3 +2473,48 @@ def test_i13307():
...
+		<p>He replied, "That's wonderful."</p>
+		<p>He replied, "That's wonderful".</p>
+		<p>He replied, "That's wonderful."[4]</p>
+		<p>Предложение по-русски.</p>
+		<p>我不会说中文！</p>
+		<p>Bye-bye, world!</p>
+	""")
+
+	expectedParagraphs = [
+		'''Hello, world!''',
```
Comment (seanbudd) [inline tests/system/robot/chromeTests.py:2502]: Shouldn't these just be single quoted strings

```suggestion
		"Hello, world!",
```

File: source/gui/settingsDialogs.py [hunk 9]
```diff
@@ -2272,6 +2272,35 @@ def makeSettings(self, settingsSizer):
...
+		# Translators: This is the label for a textfield in the
+		# browse mode settings panel.
+		textParagraphRegexLabelText = _("Regular expression for text paragraph navigation")
+		self.textParagraphRegexEdit = sHelper.addLabeledControl(
+			textParagraphRegexLabelText,
+			wxCtrlClass=wx.TextCtrl,
+			size=(self.scaleSize(300), -1),
+		)
+		self.textParagraphRegexEdit.SetValue(config.conf["virtualBuffers"]["textParagraphRegex"])
+		self.bindHelpEvent("BrowseModeSettingsTextParagraphRegexEdit ", self.textParagraphRegexEdit)
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:2284]: Removes space at end.
Also, as per [userGuideStandards](https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/userGuideStandards.md), the Feature Description Anchor should not include the settings category. Once the anchor is set it cannot be updated, while settings may move categories.

```suggestion
		self.bindHelpEvent("TextParagraphRegexEdit", self.textParagraphRegexEdit)
```

File: source/gui/settingsDialogs.py [hunk 10]
```diff
@@ -2272,6 +2272,35 @@ def makeSettings(self, settingsSizer):
...
+
+	def isValid(self) -> bool:
+		regex = self.textParagraphRegexEdit .GetValue()
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:2287]:
```suggestion
		regex = self.textParagraphRegexEdit.GetValue()
```

File: source/config/configSpec.py [hunk 11]
```diff
@@ -8,6 +8,31 @@
 from io import StringIO
 from configobj import ConfigObj
 
+DEFAULT_TEXT_PARAGRAPH_REGEX = (
```
Comment (seanbudd) [inline source/config/configSpec.py:11]: I think this should be defined elsewhere, perhaps in a sibling module named `configDefaults` or `configStrings` like `configFlags`

File: tests/system/robot/chromeTests.py (hunk 3 above)
Comment (seanbudd) [inline tests/system/robot/chromeTests.py:2507]: Could you also add comments to each of these e.g.

```suggestion
		# Tests regex condition for CJK full width character terminators
		"我不会说中文！",
```

File: user_docs/en/changes.t2t [hunk 12]
```diff
@@ -7,6 +7,7 @@ What's New in NVDA
 
 == New Features ==
 - In Windows 11, NVDA will announce alerts from voice typing and suggested actions including the top suggestion when copying data such as phone numbers to the clipboard (Windows 11 2022 Update and later). (#16009, @josephsl)
+- New QuickNav command P for jumping to next/previous text paragraph in browse mode. (#15998, @mltony)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:10]: "QuickNav" is referred to as "Quick Navigation" in NVDA core

```suggestion
- New Quick Navigation command ``p`` for jumping to next/previous text paragraph in browse mode. (#15998, @mltony)
```

File: user_docs/en/userGuide.t2t [hunk 13]
```diff
@@ -828,6 +828,7 @@ The following keys by themselves jump to the next available element, while addin
 - o: embedded object (audio and video player, application, dialog, etc.)
 - 1 to 6: headings at levels 1 to 6 respectively
 - a: annotation (comment, editor revision, etc.)
+- p: text paragraph
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:831]:
```suggestion
- ``p``: text paragraph
```

**Review by CyrilleB79 — COMMENTED** [review]
I had pending comments from a week ago; sorry, I release it only now.
Regarding my general impression on the feature, I will add it in a separate message.

**Comment by CyrilleB79** [comment]
Regarding the general experience of this feature and after testing with the add-on, I am not very comfortable using this feature. In many cases, I fear to miss some useful information. And even many times, I miss it, e.g.:
* Hheadings have usually no punctuation and are skipped
* In forums where people write quite informally and/or with spelling mistakes, I sometimes miss some short messages with no punctuation or misspelt (e.g. dot not followed by a space as usually required).

Though, I acknowledge the fact that this feature is needed. And the command to jump to a non-link block of text (N key) is probably less useful today than in the web as it was 20 years ago.

To summarize, N key catches too much text, and the current feature does not catch enough. I wonder if we could still enhance the algo to find interesting text. But I have no better suggestion, hence this comment so that any suggestion is made by other people if they have.

If no better suggestion is found, I am still in favor of merging this PR, since it seems to be an improvement for many people.

**Comment by CyrilleB79** [comment]
By the way, if the regexp for a sentence separator is validated in this PR, we should consider it as also valid for #8518 if it is implemented one day.

File: source/config/configSpec.py (hunk 11 above)
Comment (lukaszgo1) [reply source/config/configSpec.py:11]: I vote for `configDefaults` as this regex is not supposed to be used outside of the config spec.

**Comment by CyrilleB79** [comment]
Writing also here:

Regarding key assignation:
My personal opinion is that, if #16050 works as advertised, it would be much more interesting to have default assigned keys for it and keep the commands of the current PR unassigned.

Maybe we should merge this PR and #16050 with unassigned gestures and test them during alpha stage. Then at the end of alpha stage, we can figure if one or both of these feature need to have an assigned gesture and add them.

File: source/browseMode.py [hunk 14]
```diff
@@ -949,6 +995,18 @@ def _get_disableAutoPassThrough(self):
 	# Translators: Message presented when the browse mode element is not found.
 	prevError=_("no previous tab")
 )
+qn(
+	"textParagraph", key="p",
+	# Translators: Input help message for a quick navigation command in browse mode.
+	nextDoc=_("moves to the next text paragraph"),
```
[the remark this answers is not in the document]
Comment (mltony) [reply source/browseMode.py:1000]: I added a new section in the documentation explaining use cases and definition of text paragraph and also added a sentence regarding not being 100% accurate.

File: source/config/configSpec.py [hunk 15]
```diff
@@ -8,6 +8,31 @@
 from io import StringIO
 from configobj import ConfigObj
 
+DEFAULT_TEXT_PARAGRAPH_REGEX = (
+	r"({lookBehind}{optQuote}{punc}{optQuote}{optWiki}{lookAhead}|{cjk})".format(
+		# Look behind clause ensures that we have a text character before text punctuation mark.
```
[the remark this answers is not in the document]
Comment (mltony) [reply source/config/configSpec.py:13]: Am I right that in French you don't put a space before period, only before question mark and exclamation mark? If so then I adjusted regex by relaxing condition for exclamation mark and question mark. Hope that should work better.

File: source/config/configSpec.py [hunk 16]
```diff
@@ -8,6 +8,31 @@
 from io import StringIO
 from configobj import ConfigObj
 
+DEFAULT_TEXT_PARAGRAPH_REGEX = (
+	r"({lookBehind}{optQuote}{punc}{optQuote}{optWiki}{lookAhead}|{cjk})".format(
+		# Look behind clause ensures that we have a text character before text punctuation mark.
+		# We have a positive lookBehind \w that resolves to a text character in any language,
+		# coupled with negative lookBehind \d that excludes digits.
+		lookBehind=r'(?<=\w)(?<!\d)',
+		# In some cases quote or closing parenthesis might appear right before or right after text punctuation.
```
[the remark this answers is not in the document]
Comment (mltony) [reply source/config/configSpec.py:17]: Neither am I. In my understanding single quotes are typically used as apostrophe in words that are glued together, like I'm or don't and also in what's left of Genetive case in English, like Martin's - I think that's called posessive form. I can come up with a sentence that would end with apostrophe followed by period, like:

> Yesterday caucus was republicans'.
But not sure if that's even correct grammar, and if it is, then this type of sentences would be exceedingly rare, so probably not worth including into regex.

File: source/gui/settingsDialogs.py [hunk 17]
```diff
@@ -2272,6 +2272,36 @@ def makeSettings(self, settingsSizer):
 			config.conf["virtualBuffers"]["autoFocusFocusableElements"]
 		)
 
+		# Translators: This is the label for a textfield in the
+		# browse mode settings panel.
+		textParagraphRegexLabelText = _("Regular expression for text paragraph navigation")
```
[the remark this answers is not in the document]
Comment (mltony) [reply source/gui/settingsDialogs.py:2277]: Fair enough - moved this to advanced settings. Not sure who would want to change this, but it seems to me that we'd rather keep this configurable.

File: source/gui/settingsDialogs.py [hunk 18]
```diff
@@ -2272,6 +2272,36 @@ def makeSettings(self, settingsSizer):
...
 
+		# Translators: This is the label for a textfield in the
+		# browse mode settings panel.
+		textParagraphRegexLabelText = _("Regular expression for text paragraph navigation")
+		self.textParagraphRegexEdit = sHelper.addLabeledControl(
+			textParagraphRegexLabelText,
+			wxCtrlClass=wx.TextCtrl,
+			size=(self.scaleSize(300), -1),
+		)
+		self.textParagraphRegexEdit.SetValue(config.conf["virtualBuffers"]["textParagraphRegex"])
```
[the remark this answers is not in the document]
Comment (mltony) [reply source/gui/settingsDialogs.py:2283]: If a user edits NVDA config file manually - they're on their own risk. I already have a validator in `isValid()` function.

File: source/gui/settingsDialogs.py [hunk 19]
```diff
@@ -2272,6 +2272,36 @@ def makeSettings(self, settingsSizer):
...
+		self.bindHelpEvent("BrowseModeSettingsTextParagraphRegexEdit ", self.textParagraphRegexEdit)
```
[the remark this answers is not in the document]
Comment (mltony) [reply source/gui/settingsDialogs.py:2284]: Added section and updated tag name as you suggested.

**Comment by XLTechie** [comment]
With regard to "text paragraphs", in the words of the great Jack Nicholson: is 
there any other kind?

Should we really be creating a new class of paragraph, from the general way 
people already understand paragraphs, to "text paragraph", which is this new 
thing with a slightly more subtle meaning?

I too am not fond of this terminology.

What the function seems to be doing, is looking for the next "probable
paragraph". That is: a thing that sort of looks like a paragraph, even if it's 
not defined as such structurally.

It's still likely to be a paragraph, in the classical definition of a paragraph.

I'd much rather call it a "probable paragraph", or something similar that 
implies the standard definition of paragraph, and our attempt to find one.

Of course, in web/Twitter/etc. comments, it's common to misuse, not use, or 
otherwise pervert the rules of, punctuation. So this is never going to be 
entirely accurate; there will always be the potential to miss something.
I can imagine this feature morphing into an AI recognition mechanism one day 
that will be much more accurate, but that isn't happening today.

Even including at least one sentence is not required to be a paragraph (take an 
excerpted quotation, for example), but the line has to be drawn somewhere, and 
requiring a sentence is still better than the N command.

@CyrilleB79 I suspect you wouldn't need this for Word. That is, if the P command 
was implemented to jump to next paragraph in Word, I think it wouldn't need to 
be nearly so fuzzy as this. I have no idea how you would detect one at the code 
level for Word, but Word already has a good idea what a paragraph is, and there 
is likely a way to ask it. Wordpad doesn't use a virtual buffer, so doing it 
there would be different in several ways.

So while the feature might be useful in other contexts, I would bet the 
implementation would have to be radically different for each, and beyond the 
scope of this PR.

**Comment by mltony** [comment]
@CyrilleB79 

> In many cases, I fear to miss some useful information. 

That's a valid concern, however, the way I view this feature is that it allows you to quickly find the beginning of the article where you would want to start reading. Then you can use `Control+up/down` commands to read if you don't want to miss a single paragraph. TBH I don't use this command to actually read things - one example is that it skips paragraphs ending with colon - and I don't want to include colon in the regex since that creates too many false positives e.g. on wikipedia.

> By the way, if the regexp for a sentence separator is validated in this PR, we should consider it as also valid for #8518 if it is implemented one day.

For actual sentence navigation there will be a few more tweaks needed, but yes, we can keep single regexp for both use cases.

> if #16050 works as advertised, it would be much more interesting to have default assigned keys for it and keep the commands of the current PR unassigned.

Style navigation is a much more specialized command for advanced users. That's according to TextNav usage vs style navigation in BrowserNav. That's why I prefer text paragraph navigation to have a default key binding.

> Naming 

I am open to this as long as we can find a better name for it.

> And I feel that it may be easily confused with normal paragraph navigation (control+up/downArrow), which is configurable with the paragraph style parameter.

I added a sentence in the doc emphasizing the difference.

> Question 2: In NVDA there is already an algorithm using the detection of sentence end in documentNavigation/sentenceHelper.py and used for the existing paragraph navigation feature. I think that both algoritms (or at least regex) should be commonalized.

Let me prepare a PR for sentence navigation - it will come with the fanciest regex to cover most edge cases, then we can consider converging on that one.

> Question 4: Could this feature be generalized and be available not only in browse mode but also in any text? E.g. Word (browse mode off), WordPad. 

This feature should work in word in browse mode. WordPad doesn't appear to have browse mode, so maybe that's a separate question we can consider - why not bring browse mode to wordPad? Although given its limited usage it's probably not a very important issue.

File: source/browseMode.py [hunk 20]
```diff
@@ -440,9 +443,56 @@ def _iterNodesByType(self,itemType,direction="next",pos=None):
 	def _iterNotLinkBlock(self, direction="next", pos=None):
 		raise NotImplementedError
 
+	MAX_ITERATIONS_FOR_SIMILAR_PARAGRAPH = 100000
```
Comment (seanbudd) [inline source/browseMode.py:446]: Just for readability

```suggestion
	MAX_ITERATIONS_FOR_SIMILAR_PARAGRAPH = 100_000
```

File: source/config/configDefaults.py [hunk 21]
```diff
@@ -0,0 +1,33 @@
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2006-2024 NV Access Limited
```
Comment (seanbudd) [inline source/config/configDefaults.py:2]:
```suggestion
# Copyright (C) 2024 NV Access Limited
```

File: user_docs/en/userGuide.t2t [hunk 22]
```diff
@@ -842,6 +843,22 @@ If you want to use these while still being able to use your cursor keys to read
 To toggle single letter navigation on and off for the current document, press NVDA+shift+space.
 %kc:endInclude
 
++++ Text paragraph navigation command +++[TextNavigationCommand]
+
+As mentioned above, you can jupm to next or previous text paragraph by pressing ``p`` or ``shift+p``. Here we define text paragraph to be a paragraph that appears to be written in complete sentences. This can be useful to find the beginning of readable content on various webpages, such as:
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:848]:
```suggestion
You can jump to the next or previous text paragraph by pressing ``p`` or ``shift+p``.
Text paragraphs are defined by a group of text that appears to be written in complete sentences.
This can be useful to find the beginning of readable content on various webpages, such as:
```

File: user_docs/en/userGuide.t2t [hunk 23]
```diff
@@ -842,6 +843,22 @@ If you want to use these while still being able to use your cursor keys to read
...
+- Blog posts
+- 
+
+These commands can also be helpful for skipping certain kinds of clutter, such as:
+- Ads
+- Menus
+- Headers
+- 
+
+Please note, however, that while NVDA tries its best to identify text paragraphs, the algorithm is not perfect and at times can make mistakes. Please also note that this command is different from paragraph navigation commands ``control+downArrow/upArrow`` - text paragraph navigation only jumps between text paragraphs, while paragraph navigation commands take cursor to previous/next paragraphs regardless of whether they contain text or not.
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:860]:
```suggestion
Please note, however, that while NVDA tries its best to identify text paragraphs, the algorithm is not perfect and at times can make mistakes.
Additionally, this command is different from paragraph navigation commands ``control+downArrow/upArrow``.
Text paragraph navigation only jumps between text paragraphs, while paragraph navigation commands take the cursor to the previous/next paragraphs regardless of whether they contain text or not.
```

File: user_docs/en/userGuide.t2t [hunk 24]
```diff
@@ -2566,6 +2583,10 @@ This option allows you to specify if NVDA will play an error sound in case an er
 Choosing Only in test versions (default) makes NVDA play error sounds only if the current NVDA version is a test version (alpha, beta or run from source).
 Choosing Yes allows to enable error sounds whatever your current NVDA version is.
 
+==== Regular expression for text paragraph quick navigation commands ====[TextParagraphRegexEdit]
+
+This field allows users to customize regular expression for detecting text paragraphs in browse mode. [Text navigation command#TextNavigationCommand] would search for paragraphs matched by this regular expression.
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:2588]:
```suggestion
This field allows users to customize regular expression for detecting text paragraphs in browse mode.
The [text paragraph navigation command #TextNavigationCommand] searches for paragraphs matched by this regular expression.
```

**Review by seanbudd — APPROVED** [review]
Thanks @mltony

**Review by Qchristensen — APPROVED** [review]
Looks interesting and useful, thanks!

**Comment by XLTechie** [comment]
@mltony wrote:

> I wish OpenAI could create a free account for NVAccess - this way we'd be able to bring these features to all users without pushing them to set up their developer accounts

Yes. They've done something like it for Be My Eyes, so I don't imagine it's 
impossible if somebody at NV Access wanted to pursue it.

> running LLM locally - but I haven't looked much into this, I suspect it's going to be too slow for a typical consumer laptop without GPU acceleration.

Maybe, but maybe not. There are downloadable open source LLMs out there you can 
find on GitHub. I don't know that the problem is going to be the CPU/GPU speed 
so much, but also the download size is a big consideration. You need the 
training data. The last one I saw was 4 gig.

However, with the new Open AI store concept, maybe there is some way to create a 
customized LLM with limited training data on what web pages look like. This is 
not my specialty for certain.

Anyway, people don't like comments on closed PRs, so I don't want to continue 
this further, but I thought you deserved an answer.

**Comment by CyrilleB79** [comment]
Again, I have forgotten to publish pending comments on this PR...

Here are they:

1. Use "end of sentence" instead of "text paragraph navigation", i.e. the mean instead of the feature for which it is used:
   * in advanced settings: label "Regular expression for text paragraph navigation" -> "Regular expression for sentence end detection"; the User Guide may indicate what this regex is used for, i.e. text paragraph navigation, sentence navigation, etc.
   * in the code: `punctuationMarksRegex` -> `endOfSentenceRegex`
   * in the config: "textParagraphRegex" -> "endOfSentenceRegex"
   * etc.

It's clearer to read the code 

2. I hade asked:
> In NVDA there is already an algorithm using the detection of sentence end in `documentNavigation/sentenceHelper.py` and used for the existing paragraph navigation feature. I think that both algoritms (or at least regex) should be commonalized.

This factorization has not been done in this PR.

3. "browse mode settings panel" instead of "advanced settings panel" mentioned in a translator's comment
@mltony, I have not followed all and do not remember. Do you plan to implement sentence navigation soon, or is there something blocking (other than your availability of course)?
If you implement sentence navigation for NVDA 2024.2, fixing these issues may fit in the sentence navigation PR. If not, maybe it would be worth fixing this for NVDA 2024.2, especially for the name in the config, to avoid compatibility issue if we change it in the future.

**Comment by mltony** [comment]
@CyrilleB79,
I plan to work on sentence navigation in the next couple of weeks. I'll address these comments then.
